File: ts-sdk/src/blocks/get.block.with.transactions.ts
import { JsonRpcProvider, toQuantity } from "ethers";
import { Logger, WithRetry, assertExists } from "../utils";
import { Cache } from "../cache";
import { JsonRpcBlock } from "./block";

// returns a block as provided by the "eth_getBlockByNumber" or "eth_getBlockByHash" json-rpc method
export type GetBlockWithTransactions = (
  chainId: number,
  blockHashOrNumber: string | number,
  provider: JsonRpcProvider
) => Promise<JsonRpcBlock>;

export function provideGetBlockWithTransactions(
  cache: Cache,
  withRetry: WithRetry,
  logger: Logger
): GetBlockWithTransactions {
  assertExists(cache, "cache");
  assertExists(withRetry, "withRetry");
  assertExists(logger, "logger");

  return async function getBlockWithTransactions(
    chainId: number,
    blockHashOrNumber: string | number,
    provider: JsonRpcProvider
  ) {
    // check the cache first
    const cachedBlock = await cache.getBlockWithTransactions(
      chainId,
      blockHashOrNumber
    );
    if (cachedBlock) {
      // logger.debug(`cachedBlock=${cachedBlock}`);
      return cachedBlock;
    }

    // determine whether to call getBlockByNumber or getBlockByHash based on input
    let methodName = "eth_getBlockByNumber";
    if (typeof blockHashOrNumber === "string") {
      if (!blockHashOrNumber.startsWith("0x")) {
        blockHashOrNumber = parseInt(blockHashOrNumber);
      } else {
        methodName = "eth_getBlockByHash";
      }
    }

    logger.debug(
      `falling back to bot's provider for ${methodName} for chain ${chainId}...`
    );
    // fetch the block
    const block: JsonRpcBlock = await withRetry(provider.send.bind(provider), [
      methodName,
      [toQuantity(blockHashOrNumber), true],
    ]);

    if (block) {
      await cache.setBlockWithTransactions(chainId, block);
    }
    return block;
  };
}

File: py-sdk/src/forta_bot_sdk/blocks/get_block_with_transactions.py
from web3 import AsyncWeb3
from typing import Callable, Optional
from ..cache import Cache
from ..utils import assert_exists, WithRetry
from .block import Block


GetBlockWithTransactions = Callable[[
    int, str | int, AsyncWeb3], Optional[Block]]


def provide_get_block_with_transactions(cache: Cache, with_retry: WithRetry) -> GetBlockWithTransactions:
    assert_exists(cache, 'cache')
    assert_exists(with_retry, 'with_retry')

    async def get_block_with_transactions(chain_id: int, block_hash_or_number: str | int, provider: AsyncWeb3) -> Optional[Block]:
        # check cache first
        cached_block = await cache.get_block_with_transactions(chain_id, block_hash_or_number)
        if cached_block:
            return Block(cached_block)

        # determine which method to use based on input
        method_name = "eth_getBlockByNumber"
        if type(block_hash_or_number) == str:
            if not block_hash_or_number.startswith('0x'):
                block_hash_or_number = int(block_hash_or_number, 0)
            else:
                method_name = "eth_getBlockByHash"

        # fetch the block with full transactions
        response = await with_retry(provider.provider.make_request, method_name, [hex(block_hash_or_number), True])

        # if no valid block found for block hash/number
        if not (response and 'result' in response):
            return None

        block = Block(response['result'])
        # write to cache
        await cache.set_block_with_transactions(chain_id, block.to_json())

        return block

    return get_block_with_transactions
